text/0055-signature-based-finality.md

```diff
@@ -0,0 +1,115 @@
+# Signature Based Finality
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0055](https://github.com/casperlabs/ceps/pull/55)
+
+This document is a follow-up to [a point in CEP 36](https://github.com/casper-network/ceps/blob/master/text/0036-consensus-memory-usage.md#finality-is-determined-by-finality-signatures-only), specifying the details of how block finality based on signatures should be implemented.
+
+## Motivation
+
+[motivation]: #motivation
+
+The consensus component is currently quite heavy on memory, having to store the full protocol states of multiple eras from the past, in case some nodes decide to equivocate in those eras and have to be punished. Also, communicating the protocol state to other nodes is currently the only way to make them finalize blocks and advance to later eras if they are lagging behind. Thus, if we want to be able to drop older eras, we have to make sure that:
+
+- we can detect attempts to create a fork in the chain by equivocating, and
+- we can prove finality of the latest blocks to lagging nodes.
+
+Both of these goals can be achieved with signature based finality.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The current process of block finalization is this:
+
+1. A `BlockPayload` gets finalized by a consensus instance.
+2. The consensus component announces a `FinalizedBlock`.
+3. Contract runtime executes the block and announces a `Block`.
+4. Linear chain component stores the block and announces that it was stored.
+5. Consensus creates a finality signature for the block.
+6. Linear chain stores the signature and gossips it.
+
+In order to only consider fully signed blocks as finalized, we'll have to change this process to the following:
+
+1. A `BlockPayload` gets finalized by a consensus instance.
+2. The consensus component announces a `FinalizedBlock`.
+3. Contract runtime executes the block and announces a `Block`.
+4. In response to the announcement, linear chain component caches the block, and consensus creates a finality signature.
+5. Linear chain stores the signature and gossips it. It also receives signatures sent over the network.
+6. Once the linear chain gathers enough signatures for the block, it considers it finalized.
+    - If the block is unknown yet, it requests the block from the sender of the last signature.
+    - If the block is known, but it isn't the immediate descendant of the highest block so far, it requests the range of blocks from the immediate descendant of the highest block to the just finalized block.
```

Comment:
I'm guessing that the _the highest block so far_ means highest _known_ and _stored_ and _finalized_?

Reply:
Who does it request this range of blocks from? The sender of the last signature as well?

Reply:
Highest stored in the linear chain. After this proposal, it will only get stored once it is known, finalized by consensus and fully signed.

I'll clarify in the text.